README.md

```diff
@@ -1,34 +1,741 @@
+* [What is a Sum Type?](#what-is-a-sum-type)
+* [Quick Start](#quick-start)
+* [Defining Sum Types](#defining-sum-types)
+* [Predicates](#predicates)
+* [Pattern Matching with Hashes](#pattern-matching-with-hashes)
+* [Pattern Matching with Blocks](#pattern-matching-with-blocks)
+* [Methods on Sum Types](#methods-on-sum-types)
+* [Variant Instance Methods](#variant-instance-methods)
+* [A Note on Mutability](#a-note-on-mutability)
+* [Maybes](#maybes)
+* [Results](#results)
+* [Development](#development)
+* [Contributing](#contributing)
+* [License](#license)
+* [Code of Conduct](#code-of-conduct)
+
+## What is a Sum Type?
+
+Sum types are data structures with multiple variants.
+Ruby does not have sum types, but many concepts in the language (like booleans, integers, errors, etc.) can be described using sum types.
+Sum types are not limited to those use-cases, however, and are a powerful tool for modeling domain-specific data as well.
+
+This README will use non-generalized examples of sum types to help build an intuition for when they might be useful.
+To learn more about sum types, I recommend watching [Philip Wadler's Category Theory for the Working Hacker](https://www.youtube.com/watch?v=V10hzjgoklA) and checking out [Elm's Custom Types](https://guide.elm-lang.org/types/custom_types.html), [Haskell's Sum Types](https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/sum-types), and [Rust's Enums](https://doc.rust-lang.org/book/ch06-01-defining-an-enum.html).
+
+## Quick Start
+
+Define a sum type:
+
+```ruby
+Direction = SumsUp.define(:north, :south, :east, :west)
+# => Direction
+
+Direction.north
+# => #<variant Direction::North>
+
+Direction.south
+# => #<variant Direction::South>
+
+Direction.east
+# => #<variant Direction::East>
+
+Direction.west
+# => #<variant Direction::West>
+```
+
+Use predicates to distinguish between variants:
+
+```ruby
+def latitudinal?(direction)
+  direction.north? ||
+    direction.south?
+end
+
+latitudinal?(Direction.south)
+# => true
+
+latitudinal?(Direction.west)
+# => false
+```
+
+Call `#match` to categorically handle each variant by name:
+
+```ruby
+def turn_clockwise(direction)
+  direction.match do |m|
+    m.north { Direction.east }
+    m.south { Direction.west }
+    m.east { Direction.south }
+    m.west { Direction.north }
+  end
+end
+
+turn_clockwise(Direction.north)
+# => #<variant Direction::East>
+
+turn_clockwise(turn_clockwise(Direction.north))
+# => #<variant Direction::South>
+```
+
+## Defining Sum Types
+
+Imagine we're writing software for a coffee shop.
+The menu might look something like this:
+
+| Item                 | Small | Large |
+|----------------------|-------|-------|
+| Water                | Free  |       |
+| Lemonade             | $3.50 | $4.50 |
+| Coffee (Hot or Iced) | $2.95 | $3.95 |
+
+To model the menu using sum types, let's start out with some simple enumerations:
+
+```ruby
+Size = SumsUp.define(:small, :large)
```

Comment:
I wonder if there's an opportunity here to talk about how easily extensible sum types are. This exists as a 2 variant set today but we may want to introduce a medium variant in the future. A sum type ensures we have to account for that variant throughout our implementation.

I often get asked why one might choose a sum type over a boolean - and actually maybe worth giving more weight to why this is also a step above basic enums which don't force handling of all cases. 

You allude to these things in your examples but you don't call it out explicitly in the overarching benefit of the library.

Reply:
Yeah, that's a really good call. I think that would also help get at the _why_.